misc/makefile_help.go

```diff
@@ -0,0 +1,381 @@
+//	-h, --help                Show usage and exit
+package main
+
+import (
+	"bufio"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"unicode"
+)
+
+var (
+	numPrefixRE        = regexp.MustCompile(`^(\d+):`)
+	percentNumPrefixRE = regexp.MustCompile(`^%(\d+):`)
+)
+
+// flagList accumulates multiple flag values into a slice.
+type flagList []string
+
+// String returns the comma‑separated flag values.
+func (f *flagList) String() string {
+	var b strings.Builder
+	b.WriteByte('[')
+
+	for i, s := range *f {
+		if i > 0 {
+			b.WriteString(", ")
+		}
+		b.WriteString(strconv.Quote(s))
+	}
+
+	b.WriteByte(']')
+	return b.String()
+}
+
+// Set appends a new value to the flag list.
+func (f *flagList) Set(val string) error {
+	*f = append(*f, val)
+	return nil
+}
+
+// flagList accumulates multiple flag values into a slice.
+type nestedFlagList [][]string
+
+// String returns the comma‑separated flag values.
+func (nf *nestedFlagList) String() string {
+	var b strings.Builder
+	b.WriteByte('[')
+
+	for i, inner := range *nf {
+		if i > 0 {
+			b.WriteString(", ")
+		}
+		b.WriteByte('[')
+		for j, s := range inner {
+			if j > 0 {
+				b.WriteString(", ")
+			}
+			b.WriteString(strconv.Quote(s))
+		}
+		b.WriteByte(']')
+	}
+
+	b.WriteByte(']')
+	return b.String()
+}
+
+// Set appends a new value to the flag list.
+func (nf *nestedFlagList) Set(val string) error {
+	targetIdx := 0
+	rest := val
+
+	if m := numPrefixRE.FindStringSubmatch(val); m != nil {
+		idx, err := strconv.Atoi(m[1])
+		if err != nil || idx <= 0 {
+			// This should not be reachable.
+			return fmt.Errorf("invalid prefix index in %q", val)
+		}
+		targetIdx = idx - 1
+		rest = val[len(m[0]):]
+	}
+
+	// Ensure enough inner slices
+	for len(*nf) <= targetIdx {
+		*nf = append(*nf, []string{})
+	}
+
+	(*nf)[targetIdx] = append((*nf)[targetIdx], rest)
+	return nil
+}
+
+// Config holds all command‑line options.
+type Config struct {
+	Makefile   string         // path to the main Makefile
+	RelativeTo string         // base path for printing sub‑directory commands
+	Dirs       flagList       // additional directories to scan
+	Wildcards  nestedFlagList // wildcard values to expand '%' targets
+}
+
+// parseConfig parses flags and the single required Makefile argument.
+// Returns a non‑nil *Config on success, or an error on bad input.
+func parseConfig(args []string) (*Config, *flag.FlagSet, error) {
+	cfg := &Config{}
+	fs := flag.NewFlagSet("makefile-help", flag.ContinueOnError)
+
+	fs.StringVar(&cfg.RelativeTo, "relative-to", "", "base path for sub‑directory commands")
+	fs.StringVar(&cfg.RelativeTo, "r", "", "shorthand for --relative-to")
+	fs.Var(&cfg.Dirs, "dir", "directory to scan for Makefiles (repeatable)")
+	fs.Var(&cfg.Dirs, "d", "shorthand for --dir")
+	fs.Var(&cfg.Wildcards, "wildcard", "value to substitute for '%' in targets (repeatable)")
+	fs.Var(&cfg.Wildcards, "w", "shorthand for --wildcard")
```

Comment:
remove the shorthands (nobody will use this directly except for testing; better to be clear in our scripts)